### PR TITLE
feat: implement new address sorting algorithm

### LINKED ISSUE #9725: Kubelet/etcd uses wrong IPv6 Address
### Description

When Talos is run in an IPv6 Single-Stack environment and is assigned multiple IPs by DHCP and RA (although this will most likely apply to Dual-Stack as well) the Kubelet will use the wrong Address.

In our case Talos is running in KubeVirt with the Passt network binding plugin and gets an IP via RA followed by an /128 IP from DHCPv6. Only the latter has full bi-directional connectivity.

The preferred  /128 address has the flag `permanent` while the RA address has the flag `mngmtmpaddr`.

The `permanent` address should be preferred.

### Logs

Relevant excerpts from `omnictl support`:

AddressStatuses
```yaml
# cat fd01:cafe::5054:ff:fe1f:c7bd/resources/addressstatuses.net.talos.dev.yaml
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: eth0/fd01:cafe::5054:ff:fe1f:c7bd/64
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:53Z
    updated: 2024-11-14T15:52:53Z
spec:
    address: fd01:cafe::5054:ff:fe1f:c7bd/64
    linkIndex: 8
    linkName: eth0
    family: inet6
    scope: global
    flags: mngmtmpaddr
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: eth0/fd01:cafe::f14c:9fa1:8496:557f/128
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    address: fd01:cafe::f14c:9fa1:8496:557f/128
    linkIndex: 8
    linkName: eth0
    family: inet6
    scope: global
    flags: permanent
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: eth0/fe80::5054:ff:fe1f:c7bd/64
    version: 2
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:50Z
    updated: 2024-11-14T15:52:52Z
spec:
    address: fe80::5054:ff:fe1f:c7bd/64
    linkIndex: 8
    linkName: eth0
    family: inet6
    scope: link
    flags: permanent
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: lo/127.0.0.1/8
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:48Z
    updated: 2024-11-14T15:52:48Z
spec:
    address: 127.0.0.1/8
    linkIndex: 1
    linkName: lo
    family: inet4
    scope: host
    flags: permanent
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: lo/169.254.116.108/32
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    address: 169.254.116.108/32
    linkIndex: 1
    linkName: lo
    family: inet4
    scope: host
    flags: permanent
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: lo/::1/128
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:49Z
    updated: 2024-11-14T15:52:49Z
spec:
    address: ::1/128
    linkIndex: 1
    linkName: lo
    family: inet6
    scope: host
    flags: permanent
---
metadata:
    namespace: network
    type: AddressStatuses.net.talos.dev
    id: siderolink/fdae:41e4:649b:9303:2972:b262:4fad:b458/64
    version: 1
    owner: network.AddressStatusController
    phase: running
    created: 2024-11-14T15:52:53Z
    updated: 2024-11-14T15:52:53Z
spec:
    address: fdae:41e4:649b:9303:2972:b262:4fad:b458/64
    linkIndex: 9
    linkName: siderolink
    family: inet6
    scope: global
    flags: permanent
```

NodeAddresses:
```yaml
# cat fd01:cafe::5054:ff:fe1f:c7bd/resources/nodeaddresses.net.talos.dev.yaml
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: accumulative
    version: 4
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:47Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
        - fdae:41e4:649b:9303:2972:b262:4fad:b458/64
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: accumulative-no-k8s
    version: 2
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
        - fdae:41e4:649b:9303:2972:b262:4fad:b458/64
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: accumulative-only-k8s
    version: 1
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses: []
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: current
    version: 4
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:47Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
        - fdae:41e4:649b:9303:2972:b262:4fad:b458/64
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: current-no-k8s
    version: 2
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
        - fdae:41e4:649b:9303:2972:b262:4fad:b458/64
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: current-only-k8s
    version: 1
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses: []
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: default
    version: 1
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:53Z
    updated: 2024-11-14T15:52:53Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: routed
    version: 3
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:47Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: routed-no-k8s
    version: 2
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd/64
        - fd01:cafe::f14c:9fa1:8496:557f/128
---
metadata:
    namespace: network
    type: NodeAddresses.net.talos.dev
    id: routed-only-k8s
    version: 1
    owner: network.NodeAddressController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses: []
```

NodeIPs:
```yaml
# cat fd01:cafe::5054:ff:fe1f:c7bd/resources/nodeips.kubernetes.talos.dev.yaml
metadata:
    namespace: k8s
    type: NodeIPs.kubernetes.talos.dev
    id: kubelet
    version: 1
    owner: k8s.NodeIPController
    phase: running
    created: 2024-11-14T15:52:55Z
    updated: 2024-11-14T15:52:55Z
spec:
    addresses:
        - fd01:cafe::5054:ff:fe1f:c7bd
```

### Environment

- Talos version: [`talosctl version --nodes <problematic nodes>`] v1.8.2
- Kubernetes version: [`kubectl version --short`] v1.30.1
- Platform: Omni+KubeVirt

### PR DESCRIPTION
**Pull request by smira**
Fixes #9725

See #9749

### CODE REVIEW

**Comment by smira** [comment]
I'm not sure how to configure sorting algorithm (is the machine config not too late?).

Probably safe to keep old defaults.

tl;dr is trying to drop alphanumeric sort on resource IDs in favor of proper comparison which includes putting shorter prefixes first.

heavy WIP at the moment, mostly refactoring

File: internal/app/machined/pkg/controllers/network/node_address.go [hunk 1]
```diff
@@ -300,6 +308,40 @@ outer:
...
+		return -1
+	}
+
+	// compare addresses
+	if c := a.Addr().Compare(b.Addr()); c != 0 {
+		return c
+	}
+
+	// if addresses are equal, compare prefix lengths
+	return cmp.Compare(a.Bits(), b.Bits())
```
Comment (smira) [inline internal/app/machined/pkg/controllers/network/node_address.go:342]: should be `-cmp.Compare` probably

File: internal/app/machined/pkg/controllers/network/internal/addressutil/compare.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		return ComparePrefixNew
+	}
+
+	panic(fmt.Sprintf("unknown address sort algorithm: %s", algorithm))
+}
+
+// ComparePrefixesLegacy is the old way to sort prefixes.
+//
+// It only compares addresses and does not take prefix length into account.
+func ComparePrefixesLegacy(a, b netip.Prefix) int {
```
Comment (smira) [inline internal/app/machined/pkg/controllers/network/internal/addressutil/compare.go:31]: this function and the function below is the tl;dr of this change, everything else is just supporting stuff

**Review by DmitriyMV — APPROVED** [review]
